### PR TITLE
IW-1022 | global nav

### PR DESCRIPTION
**Pull request by xkxd**
This is the main PR for all the work related to implementing our current Ember-based global navigation in React.

Including:
- breaking change, described here: https://github.com/Wikia/react-common/pull/43#discussion_r260372001
- nav working with the newest version of React (16.8.3 at the moment of writing) => we use hooks, so you'll need to update it in your app in order to use that component
- tricky, yet acceptable, way of testing those hooks (eg. [here](https://github.com/Wikia/react-common/pull/22/files#diff-a996801f208b13ddb3eeeb59ac010e6aR42)), described here: https://github.com/Wikia/react-common/pull/40#discussion_r258885523
- few disabled .eslint rules, because we didn't care about them anyway

### CODE REVIEW

**Comment by vforge** [comment]
Pinging @JoshuaRogan - CAKE is using their own GN now and it will be NICE to kill that too.

File: .eslintrc [hunk 1]
```diff
@@ -54,17 +54,19 @@
     "no-restricted-imports": 0,
     "class-methods-use-this": 0,
     "no-use-before-define": 0,
+    "object-curly-newline": 0,
     "react/destructuring-assignment": 0,
     "react/no-access-state-in-setstate": "error",
     "react/jsx-filename-extension": 0,
     "react/jsx-indent": 0,
     "react/jsx-indent-props": 0,
+    "react/forbid-prop-types": 0,
```
Comment (vforge) [inline .eslintrc]: Please create `PropTypes.shape` instead of using `PropTypes.object`

File: config/styleguide.config.json [hunk 2]
```diff
@@ -51,6 +51,7 @@
       "Fieldset",
       "FloatingButton",
       "FloatingButtonGroup",
+      "GlobalNavigation",
```
Comment (vforge) [inline config/styleguide.config.json]: Please move `GlobalNavigation` to `systems/` instead`  - It's way more complex than just a component.

File: source/components/AvatarStack/index.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import React from 'react';
```
Comment (vforge) [inline source/components/AvatarStack/index.js]: Consider adding this in separate PR.

File: source/components/GlobalNavigation/hooks/useLazyLoad.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { useEffect, useRef } from 'react';
```
Comment (vforge) [inline source/components/GlobalNavigation/hooks/useLazyLoad.js]: Seems universal enough to live in `/hooks/` - please move it there and add it to the styleguide with proper instructions.

**Comment by vforge** [comment]
Still waiting on @JoshuaRogan review (as it's probably going to be used to replace custom GlobalNav in FandomCreator)

File: .eslintrc (hunk 1 above)
Comment (xkxd) [reply .eslintrc]: https://github.com/Wikia/react-common/pull/22/commits/6327fae41b68b043ce0805639660b3cc3620ebd9

File: config/styleguide.config.json (hunk 2 above)
Comment (xkxd) [reply config/styleguide.config.json]: I'd argue.

Aside of its size GlobalNav is still a single component with no additional requirements which accepts a bunch of props and does not care about the rest of app, while BannerNotification comes with store injection, actions and a dedicated model.

File: source/components/GlobalNavigation/hooks/useLazyLoad.js (hunk 4 above)
Comment (xkxd) [reply source/components/GlobalNavigation/hooks/useLazyLoad.js]: https://github.com/Wikia/react-common/pull/22/commits/8b3fabf022e2ebf2d6b673f37facc4ec760992d2

File: source/components/AvatarStack/index.js (hunk 3 above)
Comment (xkxd) [reply source/components/AvatarStack/index.js]: #49

File: source/components/AvatarStack/index.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+import PropTypes from 'prop-types';
+import classNames from 'classnames';
+
+import Avatar from '../Avatar';
+
+import './styles.scss';
+
+/* eslint-disable react/no-array-index-key */
+const AvatarStack = ({ avatars, overrideCount, maxStackSize, hideOverflow, className }) => {
```
Comment (JoshuaRogan) [inline source/components/AvatarStack/index.js]: A little late in the game now but it would be great if this were a separate PR. I would rather review three smaller contained PRs than one big one

File: config/styleguide.config.json (hunk 2 above)
Comment (JoshuaRogan) [reply config/styleguide.config.json]: I think that is a fair point but GlobalNav is more of an organism whereas components should be molecules/atoms. http://bradfrost.com/blog/post/atomic-web-design/#atoms

File: .eslintrc [hunk 6]
```diff
@@ -76,6 +77,9 @@
     "jest/valid-expect": "error",
     "react/button-has-type": 0,
     "jsx-a11y/label-has-for": 0,
+    "jsx-a11y/click-events-have-key-events": 0,
```
Comment (JoshuaRogan) [inline .eslintrc]: I'm not convinced we should disable all of these. We should try to be accessible as possible at a min in this library. We already do a poor job in most of our apps so it would be good to set the precedent in react-common.

File: source/components/AvatarStack/index.js (hunk 5 above)
Comment (xkxd) [reply source/components/AvatarStack/index.js]: It's already on master, I've just forget to update it here.

File: .eslintrc (hunk 6 above)
Comment (xkxd) [reply .eslintrc]: In general i agree, but this is a company-wide plan that should be mostly driven by UX and product folks, because it requires some specific keyboard behaviours to be introduced. I'd glad if such thing would become real, but afaik we had some approaches for it and we failed in execution. 

Having that rule for the sake of adding empty `onKeyPress` listeners to satisfy linter did not make sense, so I've decided to disable it after 5th or 6th `eslint disable next line` comment.

File: config/styleguide.config.json (hunk 2 above)
Comment (xkxd) [reply config/styleguide.config.json]: At which moment a 'component' becomes an 'organism'/'system'? It's not clear for me, and could cause some issues with understanding the concept. Eg. if we'd like to move `OpenGraph` card to common, should it be a 'component' or an 'organism'?

My understanding of the file structure was described above - if you need to export something more than a component itself, it should become a system. But in other case, from dev/react perspective, a component (no matter how big) is still a encapsulated entity that you can use and pass props to.

Maybe to avoid such issues we should rename 'systems' to eg. 'redux' and keep there only related stuff?
